go/database/mpt/io/parallel_visit.go

```diff
@@ -0,0 +1,429 @@
+// Copyright (c) 2024 Fantom Foundation
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+package io
+
+import (
+	"errors"
+	"fmt"
+	"github.com/Fantom-foundation/Carmen/go/backend/stock"
+	"github.com/Fantom-foundation/Carmen/go/backend/utils"
+	"github.com/Fantom-foundation/Carmen/go/database/mpt"
+	"github.com/Fantom-foundation/Carmen/go/database/mpt/io/heap"
+	"log"
+	"os"
+	"sync"
+	"sync/atomic"
+)
+
+//go:generate mockgen -source parallel_visit.go -destination parallel_visit_mocks.go -package io
+
+// NodeSourceFactory is a factory for NodeSource instances.
+// It provides access to nodes side to another infrastructure
+// that already accesses the nodes.
+type NodeSourceFactory interface {
+	Open() (NodeSource, error)
+}
+
+// NodeSource is a source of nodes.
+// It provides access to nodes by their ids.
+type NodeSource interface {
+	Get(mpt.NodeId) (mpt.Node, error)
+	Close() error
+}
+
+// visitAll visits all nodes in the trie rooted at the given node.
+// This method accesses nodes side to another infrastructure that may be already accessing the nodes.
+// The nodes are read multithreading, but provided via the visitor as they appear DFS in the trie.
+func visitAll(
+	sourceFactory NodeSourceFactory,
+	root mpt.NodeId,
+	visitor mpt.NodeVisitor,
+	cutAtAccounts bool,
+) error {
+
+	// The idea is to have workers processing a common queue of needed
+	// nodes sorted by their position in the depth-first traversal of the
+	// trie. The workers will fetch the nodes and put them into a shared
+	// map of nodes. The main thread will consume the nodes from the map
+	// and visit them.
+
+	type request struct {
+		position *position
+		id       mpt.NodeId
+	}
+
+	requestsMutex := sync.Mutex{}
+	requests := heap.New(func(a, b request) int {
+		return b.position.compare(a.position)
+	})
+
+	type response struct {
+		node mpt.Node
+		err  error
+	}
+	responsesMutex := sync.Mutex{}
+	responsesCond := sync.NewCond(&responsesMutex)
+	responses := map[mpt.NodeId]response{}
+
+	done := atomic.Bool{}
+	defer done.Store(true)
+
+	requests.Add(request{nil, root})
+
+	const NumWorker = 16
+
+	// open all sources for the number of workers and track possible errors
+	sources := make([]NodeSource, 0, NumWorker)
+	openingErrors := make([]error, 0, NumWorker)
+	for i := 0; i < NumWorker; i++ {
+		source, err := sourceFactory.Open()
+		sources = append(sources, source)
+		openingErrors = append(openingErrors, err)
+	}
```

Comment:
Why did you factor out the source creation from the individual workers? Creating them at the beginning of each worker seemed to be less code and complexity; also, closing them at the end of the worker's go-routine seems to be less code ...

Reply:
anyway, I changed it to open/close inside the go-routine, it needs some synchronisation to propagate errors, hope it is better, but still a bit complicated